### PR TITLE
parallel norm operation for ATen on CPU

### PR DESCRIPTION
**Pull request by xhzhao**
optimize norm operation for ATen CPU path. 
norm is a very heavy operation in RNN related workloads, see OpenNMT-py [example](https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/utils/optimizers.py#L227).
Our profiling show that norm takes about 8% in OpenNMT-py training time, which is not acceptable.

currently, the code path from TH module runs in sequential on CPU, see [link](https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THTensorMoreMath.cpp#L1937).
The norm performance data compare before and after our optimization:

norm size | SKX6148 OOB | SKX6148 OPT | speedup
-- | -- | -- | --
[35820, 500] | 19.9693 | 0.6204 | 32.19
[24997,   500] | 13.8752 | 0.4370 | 31.75
[2000, 1000] | 2.1803 | 0.0926 | 23.53
[2000,   500] | 1.0949 | 0.0508 | 21.55
[500, 1000] | 0.5483 | 0.0494 | 11.10
[500,   500] | 0.2758 | 0.0494 | 5.58

norm size | i7 OOB | i7 OPT | speedup
-- | -- | -- | --
[35820, 500] | 18.0185 | 3.8186 | 4.72
[24997,   500] | 12.5457 | 2.5238 | 4.97
[2000, 1000] | 1.8258 | 0.4204 | 4.34
[2000,   500] | 0.9252 | 0.2391 | 3.87
[500, 1000] | 0.4696 | 0.1129 | 4.16
[500,   500] | 0.2380 | 0.0650 | 3.66

### CODE REVIEW

**Comment by fmassa** [comment]
cc @colesbury, as I believe the TensorIterator reduction work can also be used for norm computation

**Comment by fmassa** [comment]
@xhzhao TensorIteration with reduction is not yet merged in master, but it will allow for more generic reductions on contiguous and non-contiguous tensors, over possibly multiple dimensions.
For a glimpse of what it will look like, have a look at https://github.com/colesbury/pytorch/tree/tensor_iterator_sum

I mentioned the TensorIterator to Sam so that we keep the `norm` operation in mind once once it is merged.

**Comment by xhzhao** [comment]
Maybe we can review this PR first, and update the norm operation after the TensorIterator PR merged.
what do you say?

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp [hunk 1]
```diff
@@ -183,9 +183,104 @@ static void prod_kernel_impl(Tensor& result, const Tensor& self, at::optional<in
...
+        int64_t n = end - begin;
+        scalar_t result = norm_calc(data, n, 1, pval);
+        return result;
+      },
+      std::plus<scalar_t>());
+    return sum;
+  }
+
+  static scalar_t norm_calc(const scalar_t* data, int64_t n, int64_t stride, float pval) {
+        scalar_t result = 0.0;
```
Comment (SsnL) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp]: It seems that the looks in this function can greatly benefit from the vectorized CPU instructions. 

Files under `native/cpu` are compiled multiple times, once for AVX2 enabled CPUs, once for AVX (but not AVX2) enabled CPUs, and once for CPUs with neither enabled. We provide helper class `Vec256` (see `vec256.h` and `vml.h`) to enabling writing the same code for different instruction sets. Other files under `native/cpu` (except `TensorCompareKernel`) have some examples. Maybe you can look into using that?

Comment (xhzhao) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp]: @SsnL yes, some case could be vectorized e.g. p=1/2/3, while some case could not, such as p=0 or non-integer. i took a look at the vec256.h and i will update my code.

Comment (fmassa) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp]: My impression is that we don't actually need to add vectorization at this point, as this code will probably be updated once TensorIterator gets in, which will significantly simplify things.

**Comment by fmassa** [comment]
Merged in #11565

**Pull request closed without merging**